core/include/detray/geometry/surfaces_finder.hpp

```diff
@@ -0,0 +1,137 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include "detray/definitions/detail/accessor.hpp"
+#include "detray/grids/axis.hpp"
+#include "detray/grids/grid2.hpp"
+#include "detray/grids/populator.hpp"
+#include "detray/grids/serializer2.hpp"
+#include "detray/tools/grid_array_helper.hpp"
+
+#pragma once
+
+namespace detray {
+
+template <std::size_t N,
+          template <typename, unsigned int> class array_type = darray,
+          template <typename...> class tuple_type = dtuple,
+          template <typename...> class vector_type = dvector,
+          template <typename...> class jagged_vector_type = djagged_vector>
+struct surfaces_finder {
+    public:
+    surfaces_finder() = delete;
+
+    using surfaces_regular_circular_grid =
+        grid2<attach_populator, axis::regular, axis::circular, serializer2,
+              vector_type, jagged_vector_type, array_type, tuple_type, dindex,
+              false>;
+
+    static const size_t N_GRIDS = N;
+
+    /// host constructor
+    template <std::size_t... ints>
+    DETRAY_HOST surfaces_finder(vecmem::memory_resource& resource)
+        : _surface_grids(host(resource, std::make_index_sequence<N_GRIDS>{})) {}
+
+    template <std::size_t... ints>
+    DETRAY_HOST auto host(vecmem::memory_resource& resource,
+                          std::index_sequence<ints...> /*seq*/) {
+
+        // TODO: Are raw pointers OK here?
+        array_type<vecmem::memory_resource*, N_GRIDS> resources;
+
+        std::fill(resources.begin(), resources.end(), &resource);
+        return darray<surfaces_regular_circular_grid, N_GRIDS>(
+            {{*resources[ints]...}});
+    }
+
+    /// device constructor
+    template <typename surfaces_finder_data_type,
+              std::enable_if_t<!std::is_base_of_v<vecmem::memory_resource,
+                                                  surfaces_finder_data_type>,
+                               bool> = true>
+    DETRAY_HOST_DEVICE surfaces_finder(surfaces_finder_data_type& finder_data)
+        : _surface_grids(
+              device(finder_data, std::make_index_sequence<N_GRIDS>{})) {}
+
+    template <std::size_t... ints, typename surfaces_finder_data_type>
+    DETRAY_HOST_DEVICE auto device(surfaces_finder_data_type& finder_data,
+                                   std::index_sequence<ints...> /*seq*/) {
+        return darray<surfaces_regular_circular_grid, N_GRIDS>(
+            {{finder_data[ints]...}});
+    }
+
+    DETRAY_HOST_DEVICE size_t size() { return N_GRIDS; }
+
+    /** Access operator - non-const
```

Comment:
Should we also maybe provide a const version from the start?

Reply:
It's fixed now